File: android/src/main/java/com/valentingrigorean/arcgis_maps_flutter/layers/FlutterServiceImageTiledLayer.java
package com.valentingrigorean.arcgis_maps_flutter.layers;

import android.net.Uri;

import com.esri.arcgisruntime.arcgisservices.TileInfo;
import com.esri.arcgisruntime.data.TileKey;
import com.esri.arcgisruntime.geometry.Envelope;
import com.esri.arcgisruntime.layers.ServiceImageTiledLayer;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;

public class FlutterServiceImageTiledLayer extends ServiceImageTiledLayer {
    private final String urlTemplate;
    private final List<String> subdomains;
    private final Map<String, String> additionalOptions;

    protected FlutterServiceImageTiledLayer(TileInfo tileInfo, Envelope fullExtent, String urlTemplate, List<String> subdomains, Map<String, String> additionalOptions) {
        super(tileInfo, fullExtent);
        this.urlTemplate = urlTemplate;
        this.subdomains = subdomains;
        this.additionalOptions = additionalOptions;
    }

    @Override
    protected String getTileUrl(TileKey tileKey) {
        final String url = urlTemplate.replace("{z}", String.valueOf(tileKey.getLevel()))
                .replace("{x}", String.valueOf(tileKey.getColumn()))
                .replace("{y}", String.valueOf(tileKey.getRow()));

        if (subdomains.size() == 0 || !url.contains("{s}")) {
            return createUrl(url);
        }

        final int subdomain = tileKey.getLevel() + tileKey.getRow() + tileKey.getColumn();

        return createUrl(url.replace("{s}", subdomains.get(subdomain % subdomains.size())));
    }

    @Override
    public String getUri() {
        return null;
    }

    private String createUrl(String url) {
        if (additionalOptions.size() == 0) {
            return url;
        }

        final Uri.Builder builder = Uri.parse(url).buildUpon();
        additionalOptions.forEach((key, value) -> {
            builder.appendQueryParameter(key, value);
        });
        return builder.build().getPath();
    }
}

File: android/src/main/java/com/valentingrigorean/arcgis_maps_flutter/layers/FlutterServiceImageTiledLayer.kt
package com.valentingrigorean.arcgis_maps_flutter.layers

import android.net.Uri
import com.arcgismaps.arcgisservices.TileKey
import com.arcgismaps.geometry.Envelope
import com.arcgismaps.mapping.layers.ServiceImageTiledLayer
import com.arcgismaps.mapping.layers.TileInfo
import com.esri.arcgisruntime.data.TileKey

class FlutterServiceImageTiledLayer(
    tileInfo: TileInfo,
    fullExtent: Envelope,
    private val urlTemplate: String,
    private val subdomains: List<String?>?,
    private val additionalOptions: Map<String?, String?>?
) : ServiceImageTiledLayer(tileInfo, fullExtent) {
    override fun getTileUrl(tileKey: TileKey): String {
        val url = urlTemplate.replace("{z}", tileKey.level.toString())
            .replace("{x}", tileKey.column.toString())
            .replace("{y}", tileKey.row.toString())
        if (subdomains!!.size == 0 || !url.contains("{s}")) {
            return createUrl(url)!!
        }
        val subdomain = tileKey.level + tileKey.row + tileKey.column
        return createUrl(url.replace("{s}", subdomains[subdomain % subdomains.size]!!))!!
    }

    override fun getUri(): String {
        return null
    }

    private fun createUrl(url: String): String? {
        if (additionalOptions!!.size == 0) {
            return url
        }
        val builder = Uri.parse(url).buildUpon()
        additionalOptions.forEach { (key: String?, value: String?) ->
            builder.appendQueryParameter(
                key,
                value
            )
        }
        return builder.build().path
    }
}